### PR TITLE
Fixed issue with textwatcher throttle invalid selection

### LINKED ISSUE #2373: Text Watcher throttling called out of sync
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open console.
1. Open manual test `/tests/plugins/autocomplete/manual/throttle`
1. Type `@anna`.
1. Remove insertion using `backspace`.

You may need to try a couple of times to get error result.

### Expected result

No errors in a console.

### Actual result

```
plugin.js?t=I6D6:7 Uncaught TypeError: Cannot read property 'element' of null
    at Object.CKEDITOR.plugins.textMatch.getRangeInText (plugin.js?t=I6D6:7)
    at Object.CKEDITOR.plugins.textMatch.match (plugin.js?t=I6D6:6)
    at b.callback (plugin.js?t=I6D6:6)
    at b.<anonymous> (plugin.js?t=I6D6:5)
    at c (ckeditor.js:27)
```

## Other details

Buffered input function for textwatcher checks is called with a already modified range.

https://github.com/ckeditor/ckeditor-dev/blob/ddff49f8b81e8495c6741f1d536de3de28dc26ba/plugins/textwatcher/plugin.js#L246-L256

* CKEditor version: 4.10.1
* Installed CKEditor plugins: textwatcher

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

Throttling textwatcher#check function sometimes results with out of sync selection. I moved selection getter into a throttled function.

Closes #2373

### CODE REVIEW

File: plugins/textwatcher/plugin.js [hunk 1]
```diff
@@ -253,7 +232,18 @@
 				return;
 			}
 
-			this._buffer.input( selectionRange );
+			var matched = this.callback( selectionRange );
```
[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/textwatcher/plugin.js]: I pushed solution based on above in the [t/2373b](https://github.com/ckeditor/ckeditor-dev/tree/t/2373b) branch. It also results with smaller LoC added compared to the original code.

[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/textwatcher/plugin.js]: Note that there's one issue - in case if during thorttled call there's no selection (edge case, but we need to consider it) there's an early return.

https://github.com/ckeditor/ckeditor-dev/blob/b5e9817/plugins/textwatcher/plugin.js#L177-L179

It would be much better to notify unmatch, because we're no longer matching anything, right? It would be nice to add unit test coverage for this edge case.

File: tests/plugins/textwatcher/textwatcher.js [hunk 2]
```diff
@@ -217,21 +224,52 @@
...
+			}, 100 );
+
+			bot.setHtmlWithSelection( '<b>[xxx]</b><i>yyy</i>' );
+
+			editable.fire( 'keyup', new CKEDITOR.dom.event( {} ) );
+			editable.fire( 'keyup', new CKEDITOR.dom.event( {} ) );
+
+			bot.setHtmlWithSelection( '<b>xxx</b><i>[yyy]</i>' );
+
+			setTimeout( function() {
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/textwatcher/textwatcher.js]: Not sure if I understand you correctly. The issue occurs when the selection changes between throttled inputs and the late input is fired with an invalid selection. 

Also, the issue is only reproducible with enabled throttling.

How could we test this case with the synchronous path? There will be no time to broke selection between `textwatcher.check` and `textwatcher._buffer.input` execution.

**Comment by jacekbogdanski** [comment]
You are right that I accidentally removed the throttling feature from `check` function. I merged branches and added required changes.

File: tests/plugins/textwatcher/textwatcher.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlewand) [reply tests/plugins/textwatcher/textwatcher.js]: What this test is doing is essentially an equivalent of the following:

```javascript
		// (#2373)
		'test throttle synchronization': function() {
			var editor = this.editor,
				bot = this.editorBot,
				textWatcher = attachTextWatcher( editor, function( selectionRange ) {
					var range = editor.getSelection().getRanges()[ 0 ];

					if ( editor.getSelection().getSelectedText() === 'yyy' ) {
						resume( function() {
							assert.isTrue( range.startContainer.equals( selectionRange.startContainer ) );
						} );
					}

					return {
						text: 'test'
					};
				}, 100 );

			bot.setHtmlWithSelection( '<b>[xxx]</b><i>yyy</i>' );

			textWatcher.check( {} );
			textWatcher.check( {} );

			bot.setHtmlWithSelection( '<b>xxx</b><i>[yyy]</i>' );

			wait();
		},
```

Note that there is no timeout. Same can be used for the next test case.

**Comment by mlewand** [comment]
I have added a couple of commits that do address some issues.
